google/resource_cloudiot_registry.go

```diff
@@ -0,0 +1,411 @@
+	eventConfigs := make([]*cloudiot.EventNotificationConfig, len(configs))
+	for i, raw := range configs {
+		data := raw.(map[string]interface{})
+
+		eventConfigs[i] = &cloudiot.EventNotificationConfig{
+			PubsubTopicName: data["pubsub_topic_name"].(string),
+		}
+	}
+	return eventConfigs
+}
+
+func expandStateNotificationConfig(config map[string]interface{}) *cloudiot.StateNotificationConfig {
+	if v, ok := config["pubsub_topic_name"]; ok {
+		return &cloudiot.StateNotificationConfig{
+			PubsubTopicName: v.(string),
+		}
+	}
+	return nil
+}
+
+func expandMqttConfig(config map[string]interface{}) *cloudiot.MqttConfig {
+	if v, ok := config["mqtt_enabled_state"]; ok {
+		return &cloudiot.MqttConfig{
+			MqttEnabledState: v.(string),
+		}
+	}
+	return nil
+}
+
+func expandHttpConfig(config map[string]interface{}) *cloudiot.HttpConfig {
+	if v, ok := config["http_enabled_state"]; ok {
+		return &cloudiot.HttpConfig{
+			HttpEnabledState: v.(string),
+		}
+	}
+	return nil
+}
+
+func expandPublicKeyCertificate(certificate map[string]interface{}) *cloudiot.PublicKeyCertificate {
+	cert := &cloudiot.PublicKeyCertificate{
+		Format:      certificate["format"].(string),
+		Certificate: certificate["certificate"].(string),
+	}
+	return cert
+}
+
+func expandCredentials(credentials []interface{}) []*cloudiot.RegistryCredential {
+	certificates := make([]*cloudiot.RegistryCredential, len(credentials))
+	for i, raw := range credentials {
+		cred := raw.(map[string]interface{})
+
+		certificates[i] = &cloudiot.RegistryCredential{
+			PublicKeyCertificate: expandPublicKeyCertificate(cred["public_key_certificate"].(map[string]interface{})),
+		}
+	}
+	return certificates
+}
+
+func expandDeviceRegistry(d *schema.ResourceData) *cloudiot.DeviceRegistry {
+	deviceRegistry := &cloudiot.DeviceRegistry{}
+	if v, ok := d.GetOk("event_notification_configs"); ok {
+		deviceRegistry.EventNotificationConfigs = expandEventNotificationConfigs(v.([]interface{}))
+	}
+	if v, ok := d.GetOk("state_notification_config"); ok {
+		deviceRegistry.StateNotificationConfig = expandStateNotificationConfig(v.(map[string]interface{}))
+
+	}
+	if v, ok := d.GetOk("mqtt_config"); ok {
+		deviceRegistry.MqttConfig = expandMqttConfig(v.(map[string]interface{}))
+
+	}
+	if v, ok := d.GetOk("http_config"); ok {
+		deviceRegistry.HttpConfig = expandHttpConfig(v.(map[string]interface{}))
+
+	}
+	if v, ok := d.GetOk("credentials"); ok {
+		deviceRegistry.Credentials = expandCredentials(v.([]interface{}))
+
+	}
+	return deviceRegistry
+}
+
+func resourceCloudiotRegistryCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region, err := getRegion(d, config)
+	if err != nil {
+		return err
+	}
+
+	parent := fmt.Sprintf("projects/%s/locations/%s", project, region)
+
+	deviceRegistry := expandDeviceRegistry(d)
```

Comment:
Here, for instance, createDeviceRegistry would be more accurate.

Reply:
👍